src/codelabs/null-safety.md

```diff
@@ -0,0 +1,336 @@
+running example code and completing exercises. To get the most out of
+this codelab, you should have some knowledge of [basic Dart syntax](/samples).
+
+{{site.alert.note}}
+  This page uses embedded DartPads to display examples and exercises.
+  {% include dartpads-embedded-troubleshooting.md %}
+{{site.alert.end}}
+
+## What is null safety?
+
+With sound null safety, types are non-nullable by default.
+That means values in your Dart programs can't be `null` unless you say they
+can be. With null safety, you can find null reference errors before you even
+run your program.
+
+The core traits of sound null safety in Dart are:
+
+-   All types are non-nullable by default.
+-   Values can't be null unless you say they can be.
+-   Finding null errors at edit time is much better than finding them at run
+    time.
+
+## Nullable and non-nullable types
+
+When you [opt in to null safety](/null-safety#enable-null-safety), all types are
+non-nullable by default. For example, if you have a variable of
+type `String`, it will always contain a string.
+
+If you want a variable of type `String` to accept any string
+or the value `null`, give the variable a nullable type by adding a
+question mark (`?`) after the type name. For example, a variable of
+type `String?` can contain a string, or it can be null.
+
+### Example: Introducing non-nullable types
+
+The variable `a` below is declared as an `int`. Try changing the value in the
+assignment to 3 or 145. Anything but null!
+
+<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/non_nullable_types.dart" replace="/145/null/g"?>
+```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
+void main() {
+  int a;
+  a = null;
+  print('a is $a.');
+}
+```
+
+### Example: Nullable types
+
+What if you need a variable that *can* hold a null value?  Try changing the
+type of `a` so that `a` can be either null or an int:
+
+<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/nullable_types.dart" replace="/int\?/int/g"?>
+```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
+void main() {
+  int a;
+  a = null;
+  print('a is $a.');
+}
+```
+
+### Example: More nullable types!
+
+Type parameters for generics can also be nullable or non-nullable. Try using
+question marks to correct the type declarations of `aNullableListOfStrings` and
+`aListOfNullableStrings`:
+
+<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/more_nullable_types.dart" replace="/String\?/String/g"?>
+```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
+void main() {
+  List<String> aListofStrings = ['one', 'two', 'three'];
+  List<String> aNullableListOfStrings = [];
+  List<String> aListofNullableStrings = ['one', null, 'three'];
+
+  print('aListofStrings is $aListofStrings.');
+  print('aNullableListOfStrings is $aNullableListOfStrings.');
+  print('aListofNullableStrings is $aListofNullableStrings.');
+}
+```
+
+## Using ? and !
```

Comment:
We've already introduced `?`, and we don't want to encourage incorrect usage of either `!` or `late`, so how about we move this section to be just before `late` (like it is in null-safe DartPad)?

```suggestion
## Using the null assertion operator (!)
```

Reply:
@ryanmcginnis I don't see this comment addressed in your changes.